### PR TITLE
DeletePrefix on the tsm1 engine

### LINKED ISSUE #1836: Teach the storage engine to apply prefix-based tombstone entries
This work relies on #1833 and #1595 being completed first.

Once we can represent prefix-based deletes via tombstone entries, we need to teach the storage engine to apply them.

### PR DESCRIPTION
**Pull request by zeebo**
This PR adds a `DeletePrefix` method to the tsm1 engine and below. It is responsible for removing all of the data for keys matching a given prefix from every tsm file, the WAL and the cache, and removing from the index any series that were fully removed due to the delete.

It is only designed to work for the "happy path" right now in that there are still race conditions if writes happen while the delete is ongoing, data structures can get out of sync on disk in the event of process failures, and there are some algorithmic inefficiencies in determining which keys to delete.

It adds a tree to keep track of which prefixes have tombstones in the index. It does this by chunking on 8-byte sections of a prefix (since most prefixes will be multiples of 8 bytes) and falling back to 1-byte sections for the rest. Rather than store it in a flat `[N]*child` like a radix tree might, it uses a `map[key]*child` making it sparse. This allows for efficient queries of the form "what prefixes match this given key?" which is used when determining the set of tombstones for a key.

Closes #1836.
Updates #1837.

### CODE REVIEW

File: tsdb/tsm1/engine_delete_prefix.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+	fs, err := e.index.RetainFileSet()
+	if err != nil {
+		return err
+	}
+	defer fs.Release()
+
+	// TODO(jeff): are the query language values still a thing?
+	// Min and max time in the engine are slightly different from the query language values.
+	if min == influxql.MinTime {
```
Comment (e-dard) [inline tsdb/tsm1/engine_delete_prefix.go]: IIRC the query-land min and max values are later/earlier than the engine ones by 1ns to allow for some special sentinal values.

File: tsdb/tsm1/reader_prefix_tree.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+func newPrefixTree() *prefixTree {
+	return &prefixTree{
+		short: make(map[byte]*prefixTree),
+		long:  make(map[prefixTreeKey]*prefixTree),
+	}
+}
+
+func (p *prefixTree) Append(prefix []byte, values ...TimeRange) {
+	if len(prefix) >= prefixTreeKeySize {
+		lookup := *(*prefixTreeKey)(unsafe.Pointer(&prefix[0]))
```
Comment (benbjohnson) [inline tsdb/tsm1/reader_prefix_tree.go]: This `unsafe` makes me a little nervous but I can't believe there's not an easier way to convert slice data to an array. I suppose you could use a `copy()`.

File: tsdb/tsm1/engine_delete_prefix.go (hunk 1 above)
Comment (e-dard) [reply tsdb/tsm1/engine_delete_prefix.go]: @zeebo correct. We don't.

File: tsdb/tsm1/reader_prefix_tree.go (hunk 2 above)
Comment (zeebo) [reply tsdb/tsm1/reader_prefix_tree.go]: Copy is always a function call whereas this little trick compiles down to a single 64bit load instruction. That being said, there's only like a 2% difference between the two, so definitely gonna drop the unsafe.

**Pull request merged**